### PR TITLE
Add point positioning by box options for point, polygon and label annotations

### PR DESCRIPTION
**Pull request by stockiNail**
This PR is adding the feature to locate the point to start drawing a point, a polygon or a label, by box options (xMin, xMax, yMin, yMax).

The radius, is not a number or equals to zero, is calculated by the minimum value between width and height of the box.

Renames the `getCircleCenterPoint` helpers function to  `getElementCenterPoint`.

Changes parameters to the `getRectCenterPoint` helpers function. Now it accepts a rect (`{x: number, y: number, width: number, height: number}`) and not an element and the flag to use the final position anymore.

### TODO

- [x] box location for point annotation
- [x] box location for polygon annotation
- [x] box location for label annotation
- [x] test cases
- [x] types
- [x] documentation
- [x] waiting for #521 approval and merge

### CODE REVIEW

File: src/helpers.js
Comment (kurkle) [inline src/helpers.js:259]: I'd call this `getElementCenterPoint`

File: types/options.d.ts
Comment (kurkle) [inline types/options.d.ts:91]: Should the coordinate types be merged?

File: src/helpers.js
Comment (stockiNail) [reply src/helpers.js:259]: Fine!

File: types/options.d.ts
Comment (stockiNail) [reply types/options.d.ts:91]: Do you mean in a unique interface? No, because box and ellipse are using only x/yMin and x/yMax and not x/yValue, bu I could extend PointCoordinates from Coordinates. Ok for you?

Comment (kurkle) [reply types/options.d.ts:91]: Was just a thought, ok to leave as is.

Comment (stockiNail) [reply types/options.d.ts]: have a look! I have change a bit with extends, therefore you don't have some many extends in the annotation options

**Review by LeeLenaleee — APPROVED** [review]

File: docs/guide/types/point.md
Comment (kurkle) [inline docs/guide/types/point.md:57]: duplicate backgrondColor

File: types/options.d.ts
Comment (kurkle) [inline types/options.d.ts:46]: Missing space between `AnnotationCoordinates` and `{`
Should take another look at the linter, it should catch this kind of stuff.

Comment (stockiNail) [reply types/options.d.ts]: added blank before {

File: docs/guide/types/point.md
Comment (stockiNail) [reply docs/guide/types/point.md]: duplicated definition has been removed

**Comment by stockiNail** [comment]
`test/fixtures/label/clip-false.js FAILED` coming from another PR. Let me have a look

**Comment by stockiNail** [comment]
npm testlog, run locally, exit code 0: https://gist.github.com/stockiNail/2245e56814b1c01233ccce02c63a756f
No clear why here it failed

File: src/types/label.js
Comment (kurkle) [inline src/types/label.js:16]: I assume you changed these for code climate? I think the old formatting was easier to read.

**Comment by stockiNail** [comment]
rerun jobs and everything went well!

File: src/helpers.js
Comment (kurkle) [inline src/helpers.js:288]: This makes me think if we should make all of the annotations be "boxes" in v2.
So each would have x,y,width,height and maybe centerX/centerY properties. 
Not sure what the benefirts would be other than consistency of the x/y between elements. Now its either top left corner or center depending of the type.

File: test/fixtures/polygon/boxLocation.js
Comment (kurkle) [inline test/fixtures/polygon/boxLocation.js:23]: I'm sure there is a good explanation, but why aren't the triangles centered vertically to the boxes?

File: src/types/label.js
Comment (stockiNail) [reply src/types/label.js:16]: Yes, I prefer the previous one as well. How can I ignore the CC issue?

Comment (kurkle) [reply src/types/label.js:16]: Cc issues are just fyi, we don't require it to pass. There is also possibility to approve the issues in cc (I've sometimes done that)

File: src/helpers.js
Comment (stockiNail) [reply src/helpers.js:288]: When I was writing this PR, I see, for instance, box and line with x2 and y2 that could be useless, having x, y, width and height.
At the end of the day, all annotations are boxes, the line as well. I think it could make sense to normalize it. It could also help for the custom one. And anyway, every annotation could add own properties if needed, of top of them.

Comment (kurkle) [reply src/helpers.js:288]: Agreed. Care to write an issue about that for V2?

File: test/fixtures/polygon/boxLocation.js
Comment (stockiNail) [reply test/fixtures/polygon/boxLocation.js:23]: let me do another test before answering just to be sure about what I'll write

Comment (kurkle) [reply test/fixtures/polygon/boxLocation.js:23]: After giving it thought, it seems obvious the center of the points is vertically centered, not the outer bounds of the triangle that I was looking at.

Comment (stockiNail) [reply test/fixtures/polygon/boxLocation.js:23]: I tested which is confirming my answer. The triangle is built using the center of the box and the circle built inside the box with radius (in this case), in this case the min between height and width of the box (because radius is NaN).

Here the result drawing the circle too:

boxes

I have changed the test case in order to highlight also the circle.

Comment (kurkle) [reply test/fixtures/polygon/boxLocation.js:23]: Good, the added circle clarifies it nicely without having to thing about it :)

File: src/helpers.js
Comment (stockiNail) [reply src/helpers.js:288]: @kurkle maybe this could be the content for new PR. What do you think?

Comment (kurkle) [reply src/helpers.js:288]: I think it would be a breaking change and we should do it for annotation plugin v2. So at this point I'd just write and issue about it so we remmeber to do it.

File: src/types/label.js
Comment (stockiNail) [reply src/types/label.js]: reverted to original format

**Review by kurkle — APPROVED** [review]

**Comment by stockiNail** [comment]
@kurkle being new as collaborator, having the approval, you gave it, can I merge by myself?
Sorry for stupid question. :(

**Comment by kurkle** [comment]
> @kurkle being new as collaborator, having the approval, you gave it, can I merge by myself? Sorry for stupid question. :(

That is not a stupid question. There are number of things to consider before merging.
 - even if approved, should this be included in next release (=yes in this case)
 - is anything pending, comments, reviews etc (=no in this case)
 - consider the order of merging, to minimize the need to rebase (no other pull requests about to be merged currently, so not needed in this case)
 - its good to wait for any running actions to finish before merging (increases changes of success)

In this case I did not merge it directly, in case you had something to add still. So yes, you can merge it 👍

**Comment by kurkle** [comment]
Btw, do you have slack? it would be easier to exchange some random thoughts in a chat.

**Comment by stockiNail** [comment]
> Btw, do you have slack? it would be easier to exchange some random thoughts in a chat.

Yes I have it but... I can not use it for this stuff.. It's a long story. Anyway Let me take time to check if and how I can use it.

**Comment by stockiNail** [comment]
> > @kurkle being new as collaborator, having the approval, you gave it, can I merge by myself? Sorry for stupid question. :(
> 
> That is not a stupid question. There are number of things to consider before merging.
> 
>     * even if approved, should this be included in next release (=yes in this case)
> 
>     * is anything pending, comments, reviews etc (=no in this case)
> 
>     * consider the order of merging, to minimize the need to rebase (no other pull requests about to be merged currently, so not needed in this case)
> 
>     * its good to wait for any running actions to finish before merging (increases changes of success)
> 
> 
> In this case I did not merge it directly, in case you had something to add still. So yes, you can merge it 👍

Thank you very much! They will be more rules for next time!

**Pull request merged**
